File: vps-renew-notification/src/tg.rs
use reqwest::StatusCode;

use crate::{Error, Result};

pub async fn send_message(token: &str, chat_id: &str, text: &str) -> Result<StatusCode> {
    let api_url = format!("https://api.telegram.org/bot{token}/sendMessage");
    let cli = reqwest::ClientBuilder::new()
        .connect_timeout(std::time::Duration::from_secs(5))
        .build()
        .map_err(Error::from)?;

    let body = [("chat_id", chat_id), ("text", text)];

    let resp = cli
        .post(&api_url)
        .form(&body)
        .send()
        .await
        .map_err(Error::from)?;

    // tracing::debug!("{:?}", resp);

    Ok(resp.status())
}

#[cfg(test)]
mod test {
    #[tokio::test]
    async fn test_bot_send_message() {
        let token = std::env::var("BOT_TOKEN").unwrap();
        let chat_id = std::env::var("BOT_CHAT_ID").unwrap();
        let code = super::send_message(&token, &chat_id, "Hello, 世界")
            .await
            .unwrap();
        println!("{code}");
    }
}

File: vps-renew-notification-go/internal/pkg/tg.go
package pkg

import (
	"fmt"
	"net/http"
	"net/url"
	"time"

	"github.com/axumrs/vps-renew-notification-go/config"
)

func TgSentBotMessage(cfg *config.BotConfig, text string) (int, error) {
	apiUrl := fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", cfg.Token)
	data := url.Values{
		"chat_id": {fmt.Sprintf("%d", cfg.ChatID)},
		"text":    {text},
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.PostForm(apiUrl, data)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()

	return resp.StatusCode, nil
}
